src/ctelnet.cpp

```diff
@@ -1622,16 +1622,23 @@ void cTelnet::processTelnetCommand(const std::string& command)
             // Using a QByteArray means there is no consideration of encoding
             // used - it is just bytes...
             QByteArray rawData = command.c_str();
+            QString transcodedData;
+            
             if (command.size() < 6) {
                 return;
             }
 
             rawData = rawData.replace(TN_BELL, QByteArray("\\\\7"));
+            
+            transcodedData = rawData;
+            transcodedData.replace(QLatin1String("\u001B"), QLatin1String("\\u001B"));
+                
+            //rawData = rawData.replace(QLatin1String("\u001B"), QLatin1String("\\u001B"));
 
-            // rawData is in the Mud Server's encoding, trim off the Telnet suboption
+            // transcodedData is in the Mud Server's encoding, trim off the Telnet suboption
             // bytes from beginning (3) and end (2):
-            rawData = rawData.mid(3, static_cast<int>(rawData.size()) - 5);
-            mpHost->mLuaInterpreter.msdp2Lua(rawData.constData());
+            transcodedData = transcodedData.mid(3, static_cast<int>(transcodedData.size()) - 5);
+            mpHost->mLuaInterpreter.msdp2Lua(transcodedData.constData());
```

Comment:
That function takes a `QByteArray` but you are trying to give it a `QString` but they are not made up from the same underlying element (`char` vs `QChar`)...

Reply:
Originally Vadi wanted me to `rawData = rawData.replace(QLatin1String("\u001B"), QLatin1String("\\u001B"));` but we discovered that the .replace doesn't like the arguments. I was just playing around on this one. I am researching the types and arguments required by .replace when used with a QByteArray now.

Reply:
The current replacement that I think you are alluding to is:
```c++
    rawData = rawData.replace(TN_BELL, QByteArray("\\\\7"));
```

That is replacing a single `char` or byte {whether it is a `signed char` or an `unsigned char` is a bit esoteric for this comment} which has the value of `'0x07'` - but which can also be written in C/C++ code as `'\a'` for the Alert character. We have `#define`d that as the macro `TN_BELL` for convenience. Now the C/C++ compiler sees the raw string `"\\\\7"` and converts each of those pairs of back slashes to a literal \ so that the replacement string now contains three bytes `"\\7"`to be put in place of the original one. That string is then fed into the Lua interpreter (which also converts pairs of back slashes into a single literal one with the result that the Lua interpreter sees a **string** - not a *character* - of `"\7"`. For the record this is a case of **[leaning tooth-pick syndrome](https://en.wikipedia.org/wiki/Leaning_toothpick_syndrome)**.

Oh, hang on, that looks like a mistake - characters outside of the range of *printable* ASCII characters are formed of a backslash followed by three digits in the range `"\000"` to `"\255"` - so that should perhaps be:
```c++
    rawData = rawData.replace(TN_BELL, QByteArray("\\\\007"));
```
Ah, testing it shows that the single digit is enough, but it could get ambiguous if there are following ASCII digits in the stream!

Well, anyhow, `\u001b` as it is written in your proposed code, technically represents a Unicode codepoint and those take more than one (8-bit) byte - to transport them over a network - they can be encoded in a number of Unicode Transport Formats (which is where **UTF-8**, **UTF-16** in both its **B**ig **E**nd and little end versions and also **UTF-32** also in **BE** and **LE** variants). At this point though we are just working on individual bytes so cannot really work on anything that operates on a higher level. As for coding things, the `QLatin1String(`...`)` wrapper can only handle Latin1 encoded characters - but that `\u` implies a UTF-8 encoding so I can understand whilst things are confusing the compiler.

Furthermore Scandum (the Guy behind the TinTin++ client and the creator of MSDP) - **has not said anything about the character encoding of content in the *Mud Server Data Protocol***! Compare that to **GMCP** which we, finally, found out has to be using UTF-8 (as it gets transported in a JSON format and that mandates that encoding system).

What you have to do is work out what you are trying to achieve, and to my mind that looks as though you are trying to convert the escape character which in ASCII and every other encoding that Mudlet handles**+** is the single byte with a value of 27 decimal (0x1b in hex) into *something* else... and seeing what you have put so far suggests to me that you probably want to do this:

```
     rawData = rawData.replace(`\x1b`, QByteArray("\\\\027"));
```

**+** not every encoding scheme out there does - go and look up **[EBCDIC](https://en.wikipedia.org/wiki/EBCDIC)** to see what I mean! :laughing:

Reply:
I do want to say your breakdown is very thorough and I am very thankful for that. Sometimes when I am hammering at a problem I can't see the answer even when it is 1 line above me. Your solution is by far more straight forward.

To add to your statement of 'What you are trying to achieve'. My goal was simply to remove, or replace ANSI ESC so that I can use the MSDP channel to transmit the vmap tiles of our mudserver to the client for use in an always on vmap UI element.

I am going to toss your line in and see where it gets.

Reply:
I can confirm that Mudlet is still able to process the data I am sending via MSDP after SlySven's suggested edit.